storage/zynq7000-sdcard/sdcard.c

```diff
@@ -0,0 +1,1268 @@
+ */
+static int sdhost_reset(sdcard_hostData_t *host, uint32_t resetType)
+{
+	uint8_t i;
+	uint32_t val;
+
+	*(host->base + SDHOST_REG_CLOCK_CONTROL) |= resetType;
+	for (i = 0; i < SDHOST_RETRIES; ++i) {
+		usleep(10);
+		val = *(host->base + SDHOST_REG_CLOCK_CONTROL) & resetType;
+		if (val == 0) {
+			return 0;
+		}
+	}
+
+	return -1;
+}
+
+
+static inline uint8_t sdcard_isWriteProtected(sdcard_hostData_t *host)
+{
+	int presState;
+	if (host->isWPPinSupported) {
+		presState = *(host->base + SDHOST_REG_PRES_STATE);
+		if ((presState & PRES_STATE_WRITE_PROT_PIN) == 0) {
+			return 1;
+		}
+	}
+
+	return 0;
+}
+
+
+static void *sdhost_mmapDMA(size_t size)
+{
+	void *p;
+	size_t sz = (size + _PAGE_SIZE - 1) & ~(_PAGE_SIZE - 1);
+
+	p = mmap(NULL, sz, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS | MAP_UNCACHED, NULL, 0);
+	if (p == MAP_FAILED) {
+		return NULL;
+	}
+
+	return p;
+}
+
+
+/* Checks physical address of a buffer in virtual memory and checks if the buffer is contiguous. */
+static addr_t sdhost_mphys(void *v, size_t *vSize)
+{
+	size_t sz;
+	addr_t phys;
+	addr_t nextPhys;
+
+	phys = va2pa(v);
+	sz = _PAGE_SIZE - (phys & (_PAGE_SIZE - 1));
+
+	while (sz < *vSize) {
+		nextPhys = va2pa(v + sz);
+		if (nextPhys != phys + sz) {
+			break;
+		}
+
+		sz += _PAGE_SIZE;
+	}
+
+	if (sz < *vSize) {
+		*vSize = sz;
+	}
+
+	return phys;
+}
+
+
+static int sdhost_allocDMA(sdcard_hostData_t *host)
+{
+	size_t psize = SDCARD_MAX_TRANSFER;
+
+	host->dmaBuffer = sdhost_mmapDMA(psize);
+	if (host->dmaBuffer == NULL) {
+		return -ENOMEM;
+	}
+
+	host->dmaBufferPhys = sdhost_mphys(host->dmaBuffer, &psize);
+
+	if (SDCARD_MAX_TRANSFER != psize) {
+		munmap(host->dmaBuffer, SDCARD_MAX_TRANSFER);
+		host->dmaBuffer = NULL;
+		return -1;
+	}
+
+	if (host->dmaBufferPhys & 3) {
+		munmap(host->dmaBuffer, SDCARD_MAX_TRANSFER);
+		host->dmaBuffer = NULL;
+		return -1;
+	}
+
+	return 0;
+}
+
+
+static int sdhost_isr(unsigned int n, void *arg)
+{
+	/* SD Host controller interrupts are triggered by level, not by edge.
+	 * This means that ISR has to remove the reason for interrupt (zero out bit in STATUS register)
+	 * or disable the interrupt (zero out bit in SIGNAL_ENABLE register) - otherwise we get stuck
+	 * in infinite interrupt loop.
+	 */
+	sdcard_hostData_t *host = (sdcard_hostData_t *)arg;
+	uint32_t activeInterrupts = *(host->base + SDHOST_REG_INTR_STATUS);
+	*(host->base + SDHOST_REG_INTR_SIGNAL_ENABLE) &= ~activeInterrupts;
+	return 0;
+}
+
+
+static int _sdio_cmdExecutionWait(sdcard_hostData_t *host, uint32_t flags, time_t wait_us)
+{
+	uint32_t val, ret = -ETIME;
+	uint32_t flagsWithErrors = flags | SDHOST_ERROR_REASONS;
+
+	int doResetCmd = 0, doResetDat = 0;
+
+	*(host->base + SDHOST_REG_INTR_SIGNAL_ENABLE) |= flagsWithErrors | SDHOST_INTR_CARD_OUT;
+	/* If there is a pending interrupt from flagsWithErrors set, it will fire now */
+	int waitret = condWait(host->eventCond, host->eventLock, wait_us);
```

Comment:
It's not optimal to execute condWait without a loop, as spurious wakeups might occur (eg. EINTR)

Reply:
The current implementation of `condWait()` [automatically waits again](https://github.com/phoenix-rtos/libphoenix/blob/bd3db90465b9ae2fab89ae72a0b8842ba7db2d13/sys/threads.c#L34) if the underlying `phCondWait` syscall returns EINTR. This is also the stated behavior in [documentation](https://github.com/phoenix-rtos/phoenix-rtos-doc/blob/master/libc/functions/c/condWait.phrtos.md) so I don't think waiting in a loop here is necessary.